Container CI refactor

ezdac opened the pull request:
This adapts the more separated service accounts for the different registry locations.

The dev build pushes to the `us-west1-docker.pkg.dev/devopsre/dev-images/op-succinct-<proposer/challenger>` registry location.

This also adds a new release-only docker build workflow that gets triggered by a github release and uses the 
`us-west1-docker.pkg.dev/devopsre/op-succinct-<proposer/challenger>` registry location.

ezdac committed:
chore(ci): switch to reusable docker build workflow

ezdac committed:
chore(ci): add docker build release workflow

This will publish the container images for proposer and challenger in
the `devopsre/op-succinct` registry folder when a semver-release is
created.

piersy commented on .github/workflows/docker-build-lite-celo-release.yaml:
```diff
@@ -0,0 +1,52 @@
+
+      - name: Build, push and scan the challenger container
+        # TODO: pin to release 3.0.0 once approved and tagged
+        uses: celo-org/reusable-workflows/.github/actions/build-container@ezdac/use-docker-metadata-action
```
Shouldn't we get the @ezdac/use-docker-metadata-action branch merged before we try to merge this PR?

ezdac replied:
Let's postpone this until we want to officially release op-succinct.

piersy commented:
What about using a reusable workflow here as the two workflows are almost the same?

Something like:
```
name: Build and push images
on:
  workflow_call:
    inputs:
      registry:
        required: true
        type: string
      service-account:
        required: true
        type: string

jobs:
  build:
    runs-on: [self-hosted, org, 8-cpu]
    permissions:
      contents: read
      id-token: write
      security-events: write
    steps:
      - uses: actions/checkout@v4
        with:
          ref: ${{ github.sha }}

      - name: Login at GCP Artifact Registry
        uses: celo-org/reusable-workflows/.github/actions/auth-gcp-artifact-registry@v2.0
        with:
          workload-id-provider: projects/1094498259535/locations/global/workloadIdentityPools/gh-op-succinct/providers/github-by-repos
          service-account: ${{ inputs.service-account }}
          docker-gcp-registries: us-west1-docker.pkg.dev

      - name: Build proposer
        uses: celo-org/reusable-workflows/.github/actions/build-container@ezdac/use-docker-metadata-action
        with:
          platforms: linux/amd64
          registry: ${{ inputs.registry }}
          app-name: ${{ github.event.repository.name }}/proposer
          context: .
          dockerfile: fault-proof/Dockerfile.proposer
          push: true

      - name: Build challenger
        uses: celo-org/reusable-workflows/.github/actions/build-container@ezdac/use-docker-metadata-action
        with:
          platforms: linux/amd64
          registry: ${{ inputs.registry }}
          app-name: ${{ github.event.repository.name }}/challenger
          context: .
          dockerfile: fault-proof/Dockerfile.challenger
          push: true
```

Parameterised like:
```
name: Push dev images
on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]

jobs:
  build:
    uses: ./.github/workflows/build-images.yml
    with:
      registry: us-west1-docker.pkg.dev/devopsre/dev-images
      service-account: op-succinct-gh-dev@devopsre.iam.gserviceaccount.com
```

ezdac committed:
chore(ci): use reusable workflow for release/dev push-container

ezdac committed:
feat(ci): add new Celo specific dockerfiles and restore upstream files

ezdac committed:
chore(ci): add to dockerignore

ezdac commented:
@piersy I changed to using a shared reusable workflow for proposer and challenger, and I also fixed the issue of the cache invalidation in my reusable workflow branch. 
Let's use that (instead of a release version) for now until we have a final release of op-succinct.

This PR https://github.com/celo-org/reusable-workflows/pull/145 also tracks the cache invalidation fix for the main-branch.

The workflow still loses connection to the worker-pod from time to time, but this is nothing that can be directly fixed here. If this persists, as a temporary fix we might have to split the proposer/challenger workflow so that individual runs are faster.

piersy commented on fault-proof/Dockerfile.challenger:
```diff
@@ -32,13 +32,12 @@ FROM rust-base AS builder
 
 # Build the challenger binary
-RUN cargo build --release --bin challenger --features eigenda
+RUN cargo build --release --bin challenger
 
 # Runtime stage (minimal image)
 FROM ubuntu:24.04
 
 WORKDIR /app
-COPY resources/ ./resources/
```
What's going on here, are you reverting this file to match upstream, to reduce our fork's overlap?

ezdac replied:
I reverted the changes in the upstream file and added celo specific docker files

piersy commented:
> @piersy I changed to using a shared reusable workflow for proposer and challenger, and I also fixed the issue of the cache invalidation in my reusable workflow branch. Let's use that (instead of a release version) for now until we have a final release of op-succinct.
> 
> This PR [celo-org/reusable-workflows#145](https://github.com/celo-org/reusable-workflows/pull/145) also tracks the cache invalidation fix for the main-branch.
> 
> The workflow still loses connection to the worker-pod from time to time, but this is nothing that can be directly fixed here. If this persists, as a temporary fix we might have to split the proposer/challenger workflow so that individual runs are faster.

The problem I see with using your branch is that when that branch gets deleted any workflows referencing it will fail. It seems like asking for trouble.

ezdac committed:
fix(ci): wrong / unicode encoding in app-name

ezdac committed:
feat(ci): upgrade to reusable-workflows v3.0.0

ezdac committed:
feat(ci): allow parallel docker build runs within one parent

ezdac committed:
chore(ci): temporarily disable trivy scan

ezdac committed:
chore(ci): increase timeouts

ezdac committed:
chore(ci): bump docker-build workflow version

ezdac committed:
ci: bump docker-build workflow version

ezdac commented:
This PR now uses a pre-release version of the reusable-workflows. We were still encountering some issues in that release, but it should be good to merge here still, because the the docker-images are building successfully and the pre-release version tag is immutable.
